frame/staking-rewards/src/lib.rs

```diff
@@ -85,11 +132,150 @@ pub mod pallet {
 		#[pallet::constant]
 		type PalletId: Get<PalletId>;
 
+		/// Maximum number of staking duration presets allowed.
+		#[pallet::constant]
+		type MaxStakingDurationPresets: Get<u32>;
+
+		/// Maximum number of reward configurations per pool.
+		#[pallet::constant]
+		type MaxRewardConfigsPerPool: Get<u32> + TypeInfo;
+
+		/// Required origin for reward pool creation.
+		type RewardPoolCreationOrigin: EnsureOrigin<Self::Origin>;
+
 		type WeightInfo: WeightInfo;
 	}
 
+	/// Abstraction over RewardPoolConfiguration type
+	type RewardPoolConfigurationOf<T> = RewardPoolConfiguration<
+		<T as frame_system::Config>::AccountId,
+		<T as Config>::MayBeAssetId,
+		<T as Config>::Balance,
+		<T as frame_system::Config>::BlockNumber,
+		StakingDurationToRewardsMultiplierConfig<<T as Config>::MaxStakingDurationPresets>,
+	>;
+
+	/// Abstraction over RewardPool type
+	type RewardPoolOf<T> = RewardPool<
+		<T as frame_system::Config>::AccountId,
+		<T as Config>::MayBeAssetId,
+		<T as Config>::Balance,
+		<T as frame_system::Config>::BlockNumber,
+		StakingDurationToRewardsMultiplierConfig<<T as Config>::MaxStakingDurationPresets>,
+		<T as Config>::MaxRewardConfigsPerPool,
+	>;
+
 	#[pallet::pallet]
-	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::generate_store(pub (super) trait Store)]
 	#[pallet::without_storage_info]
 	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	#[pallet::getter(fn pool_count)]
+	#[allow(clippy::disallowed_types)]
+	pub type RewardPoolCount<T: Config> =
+		StorageValue<_, T::PoolId, ValueQuery, Counter<ZeroInit, SafeIncrement, SafeDecrement>>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn pools)]
+	pub type RewardPools<T: Config> = StorageMap<_, Blake2_128Concat, T::PoolId, RewardPoolOf<T>>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create a new reward pool based on the config.
+		///
+		/// Emits `RewardPoolCreated` event when successful.
+		#[pallet::weight(T::WeightInfo::create_reward_pool())]
+		#[transactional]
+		pub fn create_reward_pool(
+			origin: OriginFor<T>,
+			pool_config: RewardPoolConfigurationOf<T>,
+		) -> DispatchResult {
+			T::RewardPoolCreationOrigin::ensure_origin(origin)?;
+			let (owner, pool_id, end_block) = match pool_config {
+				RewardRateBasedIncentive {
```

Comment:
Other incetives this code will be same. So not sure why split a such way

Reply:
We don't know how other incentives/logic would look for now. This just leaves us the flexibility to branch the code cleanly later

Reply:
i do not agree. it add ambiguility. you can still init without intetive settings setting end block now or setting rate to 0. it is up to UI to make such kind of sloppy enum, not this pallet. if this is needed as some wrapper than can do it off pallet or as separate helper extrinsic.

Reply:
I think this enum should not be part of main API, and should be part of layer on top.

Reply:
The api should be forward compatible so I would like this to be part of the API

Reply:
cc @KaiserKarel

Reply:
Let's go with the enum for now. We can always create anothe, better designed extrinsic and migrate towards that.

Reply:
hm. i proposed remove enum. enum here is overenineering imho. 

not sure why we handling backward compatibility on this level. never seen such pallets before. so it is also no the substrate design.

Reply:
i do not see a reason why we overcomplicate api and going agains how other extrinsics designed, and that is on top that enum creates ambiguilty.

Reply:
@KaiserKarel if you are ok please approve.